source/common/protobuf/utility.cc

```diff
@@ -506,6 +508,114 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
+  if (typed_struct != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, typed_struct->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact udpa::type::v1::TypedStruct with unknown type URL {}",
+                     typed_struct->type_url());
+      return;
+    }
+
+    MessageUtil::jsonConvert(typed_struct->value(), ProtobufMessage::getNullValidationVisitor(),
+                             *typed_message);
+    redactInPlace(typed_message.get(), ancestor_is_sensitive);
+    MessageUtil::jsonConvert(*typed_message, *(typed_struct->mutable_value()));
+    return;
+  }
+
+  // Otherwise, use reflection to traverse all populated fields of this message...
+  const auto* reflection = message->GetReflection();
+  std::vector<const Protobuf::FieldDescriptor*> field_descriptors;
+  reflection->ListFields(*message, &field_descriptors);
+  for (const auto* field_descriptor : field_descriptors) {
+    // Redact if this field or any of its ancestors have the `sensitive` option set.
+    const bool sensitive = ancestor_is_sensitive ||
+                           field_descriptor->options().GetExtension(udpa::annotations::sensitive);
+
+    if (field_descriptor->type() == Protobuf::FieldDescriptor::TYPE_MESSAGE) {
+      // Recursive case: traverse message fields.
+      if (field_descriptor->is_repeated()) {
+        const int field_size = reflection->FieldSize(*message, field_descriptor);
+        for (int i = 0; i < field_size; ++i) {
+          redactInPlace(reflection->MutableRepeatedMessage(message, field_descriptor, i),
+                        sensitive);
+        }
+      } else {
+        redactInPlace(reflection->MutableMessage(message, field_descriptor), sensitive);
+      }
+    } else if (sensitive) {
+      // Base case: replace strings with "[redacted]" and clear all others.
+      if (field_descriptor->type() == Protobuf::FieldDescriptor::TYPE_STRING) {
+        if (field_descriptor->is_repeated()) {
+          const int field_size = reflection->FieldSize(*message, field_descriptor);
+          for (int i = 0; i < field_size; ++i) {
+            reflection->SetRepeatedString(message, field_descriptor, i, "[redacted]");
+          }
+        } else {
+          reflection->SetString(message, field_descriptor, "[redacted]");
+        }
+      } else {
+        reflection->ClearField(message, field_descriptor);
+      }
+    }
+  }
+}
+
+} // namespace
+
+std::unique_ptr<Protobuf::Message> MessageUtil::redact(const Protobuf::Message& message) {
+  // 1. Clone the original message. New() will return an empty message of the correct dynamic type.
```

Comment:
slightly worried about the copying here, shall we let the caller do this? (i.e. taking Message& and redact in place.

Reply:
I can add an overload that takes `Message&`, but I'd bet money that nobody will ever call it. You almost certainly need to make a copy (either within `redact()` or externally at the call site) because your config will be broken if you redact it in place.

Reply:
Not really, all config dumps are already a copy of config, so there's no need to make an extra copy?